loopy/schedule/tools.py

```diff
@@ -138,3 +140,215 @@ def get_return_from_kernel_mapping(kernel):
+                dt, pos = map_.get_var_dict()[iname]
+                map_ = map_.project_out(dt, pos, 1)
+
+        for name, (dt, pos) in map_.get_var_dict().items():
+            if dt == isl.dim_type.in_:
+                tag, = filter_iname_tags_by_type(knl.inames[name].tags,
+                                                 HardwareConcurrentTag)
+                map_ = map_.set_dim_name(dt, pos, str(tag))
+
+        for i_l in lsize:
+            if f"l.{i_l}" not in map_.get_var_dict():
+                ndim = map_.dim(isl.dim_type.in_)
+                map_ = map_.add_dims(isl.dim_type.in_, 1)
+                map_ = map_.set_dim_name(isl.dim_type.in_, ndim, f"l.{i_l}")
+
+        for i_g in gsize:
+            if f"g.{i_g}" not in map_.get_var_dict():
+                ndim = map_.dim(isl.dim_type.in_)
+                map_ = map_.add_dims(isl.dim_type.in_, 1)
+                map_ = map_.set_dim_name(isl.dim_type.in_, ndim, f"g.{i_g}")
+
+        for pos in range(map_.dim(isl.dim_type.out)):
+            map_ = map_.set_dim_name(isl.dim_type.out, pos, f"_lp_dim{pos}")
+
+        updated_maps.append(map_)
+
+    map_a, map_b = updated_maps
+
+    # }}}
+
+    # {{{ Step 2: rename all lid's, gid's in map_a to lid.A, gid.A
+
+    for name, (dt, pos) in map_a.get_var_dict().items():
+        if dt == isl.dim_type.in_:
+            map_a = map_a.set_dim_name(dt, pos, name+".A")
+
+    # }}}
+
+    # {{{ Step 3: rename all lid's, gid's in map_b to lid.B, gid.B
+
+    for name, (dt, pos) in map_b.get_var_dict().items():
+        if dt == isl.dim_type.in_:
+            map_b = map_b.set_dim_name(dt, pos, name+".B")
+
+    # }}}
+
+    # {{{ Step 4: make map_a, map_b ISL sets
+
+    map_a, map_b = isl.align_two(map_a, map_b)
+    map_a = map_a.move_dims(isl.dim_type.in_, map_a.dim(isl.dim_type.in_),
+                            isl.dim_type.out, 0, map_a.dim(isl.dim_type.out))
+
+    map_b = map_b.move_dims(isl.dim_type.in_, map_b.dim(isl.dim_type.in_),
+                            isl.dim_type.out, 0, map_b.dim(isl.dim_type.out))
+    set_a = map_a.domain()
+    set_b = map_b.domain()
+
+    # }}}
+
+    assert set_a.get_space() == set_b.get_space()
+
+    # {{{ Step 5: create the set any(l.i.A != l.i.B) OR any(g.i.A != g.i.B)
+
+    space = set_a.space
+    unequal_global_id_set = isl.Set.empty(set_a.get_space())
+
+    for i_l in lsize:
+        lid_a = p.Variable(f"l.{i_l}.A")
+        lid_b = p.Variable(f"l.{i_l}.B")
+        unequal_global_id_set |= (isl_set_from_expr(space,
+                                                    p.Comparison(lid_a, "!=", lid_b))
+                                  )
+
+    for i_g in gsize:
+        gid_a = p.Variable(f"g.{i_g}.A")
+        gid_b = p.Variable(f"g.{i_g}.B")
+        unequal_global_id_set |= (isl_set_from_expr(space,
+                                                    p.Comparison(gid_a, "!=", gid_b))
+                                  )
```

Comment:
You can probably make that one `for` loop.

Reply:
I think we would have to reformulate in terms of `itertools.chain` leading to hard-to-read code.